File: src/main/java/no/fintlabs/ConfigurationService.java
package no.fintlabs;

import no.fintlabs.mapping.ConfigurationMappingService;
import no.fintlabs.model.configuration.dtos.ConfigurationDto;
import no.fintlabs.model.configuration.dtos.ConfigurationPatchDto;
import no.fintlabs.model.configuration.entities.Configuration;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class ConfigurationService {

    private final ConfigurationRepository configurationRepository;
    private final ConfigurationMappingService configurationMappingService;
    private final ObjectMappingRepository objectMappingRepository;

    public ConfigurationService(
            ConfigurationRepository configurationRepository,
            ConfigurationMappingService configurationMappingService,
            ObjectMappingRepository objectMappingRepository) {
        this.configurationRepository = configurationRepository;
        this.configurationMappingService = configurationMappingService;
        this.objectMappingRepository = objectMappingRepository;
    }

    public Optional<ConfigurationDto> findById(Long configurationId, boolean excludeMapping) {
        return configurationRepository
                .findById(configurationId)
                .map(configuration -> configurationMappingService.toDto(configuration, excludeMapping));
    }

    public Page<ConfigurationDto> findAll(ConfigurationFilter filter, boolean excludeMapping, Pageable pageable) {
        Configuration configurationExample = new Configuration();
        filter.getIntegrationId().ifPresent(configurationExample::setIntegrationId);
        filter.getCompleted().ifPresent(configurationExample::setCompleted);

        return configurationRepository
                .findAll(
                        Example.of(configurationExample),
                        pageable
                )
                .map(configuration -> configurationMappingService.toDto(
                        configuration,
                        excludeMapping
                ));
    }

    public ConfigurationDto save(ConfigurationDto configurationDto) {
        return configurationMappingService.toDto(
                configurationRepository.saveWithVersion(
                        configurationMappingService.toEntity(configurationDto)
                ),
                false
        );
    }

    public ConfigurationDto updateById(Long configurationId, ConfigurationPatchDto configurationPatchDto) {
        Configuration configuration = configurationRepository.findById(configurationId)
                .orElseThrow(IllegalArgumentException::new);

        configurationPatchDto.getIntegrationMetadataId().ifPresent(configuration::setIntegrationMetadataId);
        configurationPatchDto.isCompleted().filter(Boolean::booleanValue).ifPresent(configuration::setCompleted);
        configurationPatchDto.getComment().ifPresent(configuration::setComment);
        configurationPatchDto.getMapping().map(configurationMappingService::toEntity)
                .map(objectMappingRepository::save)
                .ifPresent(configuration::setMapping);

        return configurationMappingService.toDto(
                configurationRepository.saveWithVersion(configuration),
                false
        );
    }

    public void deleteById(Long configurationId) {
        configurationRepository.deleteById(configurationId);
    }

}

File: src/main/kotlin/no/novari/flyt/configuration/ConfigurationService.kt
package no.novari.flyt.configuration

import no.novari.flyt.configuration.mapping.ConfigurationMappingService
import no.novari.flyt.configuration.model.configuration.dtos.ConfigurationDto
import no.novari.flyt.configuration.model.configuration.dtos.ConfigurationPatchDto
import no.novari.flyt.configuration.model.configuration.entities.Configuration
import org.springframework.data.domain.Example
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

@Service
class ConfigurationService(
    private val configurationRepository: ConfigurationRepository,
    private val configurationMappingService: ConfigurationMappingService,
    private val objectMappingRepository: ObjectMappingRepository,
) {
    fun findById(
        configurationId: Long,
        excludeMapping: Boolean,
    ): ConfigurationDto? =
        configurationRepository.findByIdOrNull(configurationId)?.let {
            configurationMappingService.toDto(it, excludeMapping)
        }

    fun findAll(
        filter: ConfigurationFilter,
        excludeMapping: Boolean,
        pageable: Pageable,
    ): Page<ConfigurationDto> {
        val configurationExample =
            Configuration().apply {
                filter.integrationId?.let { integrationId = it }
                filter.completed?.let { isCompleted = it }
            }

        return configurationRepository
            .findAll(Example.of(configurationExample), pageable)
            .map { configuration ->
                configurationMappingService.toDto(configuration, excludeMapping)
            }
    }

    fun save(configurationDto: ConfigurationDto): ConfigurationDto =
        configurationMappingService.toDto(
            configurationRepository.saveWithVersion(configurationMappingService.toEntity(configurationDto)),
            false,
        )

    fun updateById(
        configurationId: Long,
        configurationPatchDto: ConfigurationPatchDto,
    ): ConfigurationDto {
        val configuration = requireNotNull(configurationRepository.findByIdOrNull(configurationId))

        configurationPatchDto.integrationMetadataId?.let { configuration.integrationMetadataId = it }
        configurationPatchDto.completed?.takeIf { it }?.let { configuration.isCompleted = it }
        configurationPatchDto.comment?.let { configuration.comment = it }
        configurationPatchDto.mapping
            ?.let(configurationMappingService::toEntity)
            ?.let(objectMappingRepository::save)
            ?.let { configuration.mapping = it }

        return configurationMappingService.toDto(
            configurationRepository.saveWithVersion(configuration),
            false,
        )
    }

    fun deleteById(configurationId: Long) = configurationRepository.deleteById(configurationId)
}
